# Milestone 2
# Part 1: 
- Merge 6 datasets covering digital access, agriculture output, food stability, financial inclusion, mobile money transaction, and demography in 10 ASEAN countries, and 6 South Asian countries.
- Goal: Provide a comprehensive for other researchers for their future research. The final merged dataset has over 130 variables. We thought this merged dataset would be helpful for researchers who are interested in analyzing the relationship between Fintech and agricultural poor in Asia. 

# Part 2:
- We refine our research scope for Milestone 2 and process, merge, and use only limited number variables to get descriptive statistics, visualizations, and regression analysis.


______________________________



Research Question: What is the impact of mobile cellular subscriptions on food supply in ASEAN and South Asian countries?


Empirical Model: 
$$
\log(\text{Food Supply})_{it} = \beta_0 + \beta_1 \log(\text{Mobile Subscriptions})_{it} + \beta_2 X_{it} + \epsilon_{it}
$$


**1.	Variables:**

-	Dependent Variable: Food supply quantity 

-	Independent Variables: Mobile cellular subscription

-	Control Variables: total population, urban population, rural population, life expectancy, mortality rate, population density, poverty headcount ratio, population growth, fertility rate, GDP per capita
 
-	Time: 2011-2024
-	Region: ASEAN and South Asia
-	Number of Country: 16

    -	ASEAN: 'VNM','LAO','THA','KHM','MYS','SGP','MMR','PHL','BRN','IDN'
    -	South Asia = 'BGD','IND','PAK','NPL','LKA','BTN'

-	Note: (Which variable) is collected only every 4 year

**2.	Data Source (Detailed link is listed in [Readme](https://github.com/Graspp-25-Spring/graspp_2025s_fintech/blob/review_milestone2/README.md)) and Download Method:** 

- Dependent Variable: 
    -	Source: Credit to Agriculture, Value Added (Agriculture, Forestry and Fishing), Value Added (Manufacture of food and beverages), Fertilizer consumption are from World Bank FAO; Use of Financial Services, mobile money transactions is from Financial Access Survey, IMF
    -	Method: API and manually download csv files

- Independent Variables: 
    -	World Development Indicators (WDI) and Findex
    -	Method: construct the download workbank data function to fetch data


- Control Variables: 
    -	Source: All from World Development Indicators (WDI), World Bank 
    -	Method: construct the download workbank data function to fetch data




# Part 1

# 1. Data sets

In [36]:
import pandas as pd

In [37]:
def basic_stats(df):
    countries = list(df['country'].unique())
    n_countries = len(countries) 
    time_series = f"{df.year.min()} ~ {df.year.max()}"
    missing_data_summary = df.drop(['country','year'], axis=1).notna().sum()
    
    return {
        'countries (first 5)': countries[:5],
        'country_count': n_countries,
        'time_series': time_series,
        'observations_bycolumn': missing_data_summary
    }


## 1.1 Agri (Agriculture Output)
- Source: World Bank FAO

In [38]:
df_agri = pd.read_csv('../data/processed/Agriculture_data.csv')
df_agri = df_agri.rename(columns={'code': 'country'})
df_agri = df_agri[(df_agri['year'] >= 2011) & (df_agri['year'] <= 2024)]
df_agri.head(2)

country  year   FarmCredit  ICTPolicy  ProductionValue  ProcessingValue  \
11     BGD  2011  3271.100803        NaN     1.649775e+06              NaN   
12     BGD  2012  3213.392664        NaN     1.854820e+06    249808.744409   

    Fertilizer  
11         NaN  
12         NaN

In [39]:
basic_stats(df_agri)

{'countries (first 5)': ['BGD', 'BRN', 'BTN', 'IDN', 'IND'],
 'country_count': 16,
 'time_series': '2011 ~ 2023',
 'observations_bycolumn': FarmCredit         42
 ICTPolicy          34
 ProductionValue    55
 ProcessingValue    82
 Fertilizer          0
 dtype: int64}

- In this section, we examine agricultural output using data from the World Bank FAO. The dataset includes key indicators such as farm credit, ICT policy presence, agricultural production value, food and beverage processing value, and fertilizer usage. Our summary shows data coverage for 16 countries from 2000 to 2023. 

- Among these indicators, production and processing values have the most complete records, with 99 and 111 observations respectively. This suggests relatively strong data availability for measuring economic contributions of agricultural and food sectors, while access to farm credit and ICT policy information is more limited across countries and years.

## 1.2 Con (Control Variables)
- Source: WDI

In [40]:
df_con = pd.read_csv('../data/processed/control_var.csv')
df_con = df_con.drop(['country'], axis = 'columns').rename(columns={'code': 'country'})
df_con = df_con[(df_con['year'] >= 2011) & (df_con['year'] <= 2024)]
df_con.head(2)

country  year  Fertility rate, total (births per woman)  \
51     BGD  2011                                     2.309   
52     BGD  2012                                     2.263   

    GDP per capita, PPP (constant 2021 international $)  \
51                                        4352.426434     
52                                        4592.048409     

    Life expectancy at birth, total (years)  \
51                                   68.499   
52                                   68.989   

    Mortality rate, infant (per 1,000 live births)  \
51                                            35.4   
52                                            34.1   

    Population density (people per sq. km of land area)  \
51                                        1179.926834     
52                                        1191.289091     

    Population growth (annual %)  Population, total  \
51                      0.908526        153591076.0   
52                      0.958356        155070101.0   

    Poverty headcount ratio at national poverty lines (% of population)  \
51                                                NaN                     
52                                                NaN                     

    Rural population (% of total population)  \
51                                    68.775   
52                                    68.007   

    Urban population (% of total population)  
51                                    31.225  
52                                    31.993

In [41]:
basic_stats(df_con)

{'countries (first 5)': ['BGD', 'BTN', 'BRN', 'KHM', 'IND'],
 'country_count': 16,
 'time_series': '2011 ~ 2023',
 'observations_bycolumn': Fertility rate, total (births per woman)                               208
 GDP per capita, PPP (constant 2021 international $)                    207
 Life expectancy at birth, total (years)                                208
 Mortality rate, infant (per 1,000 live births)                         208
 Population density (people per sq. km of land area)                    192
 Population growth (annual %)                                           208
 Population, total                                                      208
 Poverty headcount ratio at national poverty lines (% of population)     52
 Rural population (% of total population)                               208
 Urban population (% of total population)                               208
 dtype: int64}

- In this section, we collect 10 control variables from WDI, covering economic development, population estimatesand composition, poverty prevalence, and vital statistics. Our summary shows data coverage for 16 countries from 1960 to 2023.

- Among these indicators, except poverty headcount ratio has many missing values, all other 9 variables have a considerable amount of data to serve as background variables. 

## 1.3 GFI (Global Financial Inclusion)
- Source: Findex

In [42]:
df_gfi = pd.read_csv('../Data/processed/cleaned_output_2011_2022.csv')
df_gfi = df_gfi.rename(columns={'code': 'country'})
df_gfi = df_gfi[(df_gfi['year'] >= 2011) & (df_gfi['year'] <= 2024)]
df_gfi.head(2)

country  year  Account (% age 15+)  \
0     BGD  2011                31.74   
1     BGD  2012                31.49   

   Account ownership at a financial institution or with a mobile-money-service provider (% of population ages 15+)  \
0                                              31.74                                                                 
1                                              31.49                                                                 

   Account ownership at a financial institution or with a mobile-money-service provider, female (% of population ages 15+)  \
0                                          26.010000                                                                         
1                                          26.156667                                                                         

   Account ownership at a financial institution or with a mobile-money-service provider, male (% of population ages 15+)  \
0                                              37.29                                                                       
1                                              36.65                                                                       

   Account ownership at a financial institution or with a mobile-money-service provider, older adults (% of population ages 25+)  \
0                                              36.76                                                                               
1                                              36.32                                                                               

   Account ownership at a financial institution or with a mobile-money-service provider, poorest 40% (% of population ages 15+)  \
0                                          19.060000                                                                              
1                                          20.383333                                                                              

   Account ownership at a financial institution or with a mobile-money-service provider, primary education or less (% of population ages 15+)  \
0                                              21.06                                                                                            
1                                              22.34                                                                                            

   Account ownership at a financial institution or with a mobile-money-service provider, richest 60% (% of population ages 15+)  \
0                                          40.180000                                                                              
1                                          38.883333                                                                              

   ...  \
0  ...   
1  ...   

   Used a mobile phone or the internet to check account balance(% with a financial institution account, age 15+)  \
0                                              10.63                                                               
1                                              10.63                                                               

   Used a mobile phone or the internet to pay bills, female (% age 15+)  \
0                                               1.54                      
1                                               1.54                      

   Used a mobile phone or the internet to pay bills, male (% age 15+)  \
0                                               3.36                    
1                                               3.36                    

   Used a mobile phone or the internet to pay bills, rural (% age 15+)  \
0                                               8.75                     
1                                               8.75                     

   Used a mobile phone or the internet to pay bills, urban (% age 15+)  \
0                             

In [43]:
basic_stats(df_gfi)

{'countries (first 5)': ['BGD', 'BRN', 'BTN', 'IDN', 'IND'],
 'country_count': 17,
 'time_series': '2011 ~ 2022',
 'observations_bycolumn': Account (% age 15+)                                                                                                              192
 Account ownership at a financial institution or with a mobile-money-service provider (% of population ages 15+)                  192
 Account ownership at a financial institution or with a mobile-money-service provider, female (% of population ages 15+)          192
 Account ownership at a financial institution or with a mobile-money-service provider, male (% of population ages 15+)            192
 Account ownership at a financial institution or with a mobile-money-service provider, older adults (% of population ages 25+)    192
                                                                                                                                 ... 
 Used a mobile phone or the internet to send money (% age

For the Global Financial Inclusion (GFI) dataset from Findex, we collect account ownership and digital financial service usage across 17 countries from 2011 to 2022. The dataset includes over 100 indicators detailing ownership by demographic breakdowns (e.g., gender, income, education), as well as behaviors such as using mobile phones or the internet to send money.

## 1.4 Staple (Staple Food Output)
- Source: FAO

In [44]:
df_staple = pd.read_csv('../Data/processed/StapleFoodStability_adjusted.csv')
df_staple = df_staple.rename(columns={'code': 'country'})
df_staple=df_staple.rename(columns={'rolling_std': 'foodSupply_stability'})
df_staple = df_staple[(df_staple['year'] >= 2011) & (df_staple['year'] <= 2024)]
df_staple.head(2)

country  year  Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer  \
1     AFG  2011                                             197.29              
2     AFG  2012                                             190.31              

   Food supply quantity (kg/capita/yr)_Starchy Roots  foodSupply_stability  
1                                               5.72                   NaN  
2                                               6.40                   NaN

In [45]:
basic_stats(df_staple)

{'countries (first 5)': ['AFG', 'BGD', 'BTN', 'KHM', 'IND'],
 'country_count': 18,
 'time_series': '2011 ~ 2022',
 'observations_bycolumn': Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer    208
 Food supply quantity (kg/capita/yr)_Starchy Roots               208
 foodSupply_stability                                            205
 dtype: int64}

For the staple food dataset, we collect 18 countries from 2010-2022 over three indicators, food supply quantity for cereals, food supply quantity for starch, standard deviations. 

## 1.5 Access (to digital finance)
- Source: WDI

In [46]:
df_access = pd.read_csv('../Data/processed/access_merged.csv')
df_access['year'] = pd.to_datetime(df_access['year']).dt.year
df_access = df_access.rename(columns={'code': 'country'})
df_access = df_access[(df_access['year'] >= 2011) & (df_access['year'] <= 2024)]
df_access.head(2)


country  year  Rural Access to Electricity(Percent of Population)  \
0     BGD  2024                                                NaN    
1     BGD  2023                                               99.6    

   Mobile Cellular Subscriptions (per 100 people)  \
0                                             NaN   
1                                             NaN   

   Fixed Broadband Subsciptions (per 100 people)  
0                                            NaN  
1                                        7.51447

In [47]:
basic_stats(df_access)

{'countries (first 5)': ['BGD', 'BTN', 'IND', 'LKA', 'NPL'],
 'country_count': 16,
 'time_series': '2011 ~ 2024',
 'observations_bycolumn': Rural Access to Electricity(Percent of Population)    204
 Mobile Cellular Subscriptions (per 100 people)        190
 Fixed Broadband Subsciptions (per 100 people)         201
 dtype: int64}

For the access to digital finance from WDI, we collect 16 countries from 1960-2024, covering variables like rural access to electricity, mobile subscriptions, and fixed broadband subscriptions. Among which, mobile sucscription has the largest amount of available data for us to use.

## 1.6 Mobile (Mobile Money Transactions)
- Source: Financial Access Survey, IMF

In [48]:
df_mobile = pd.read_csv('../Data/processed/mobile_transact.csv')
# Filter for specific countries in South Asia and Southeast Asia
country_list = ['AFG', 'BGD', 'BRN', 'BTN', 'IDN', 'IND', 'IRN', 'KHM', 'LAO',
       'LKA', 'MDV', 'MMR', 'MYS', 'NPL', 'PAK', 'PHL', 'SGP', 'THA',
       'TLS', 'VNM', 'WLD']
df_mobile = df_mobile[df_mobile['code'].isin(country_list)]

df_mobile = df_mobile.rename(columns={'code': 'country'})
df_mobile = df_mobile[['country', 'year', 'mobile_money_transactions']]
df_mobile = df_mobile[(df_mobile['year'] >= 2011) & (df_mobile['year'] <= 2024)]
df_mobile.head(2)


country  year  mobile_money_transactions
80     IDN  2016               7.063689e+06
81     IDN  2017               1.237547e+07

In [49]:
basic_stats(df_mobile)

{'countries (first 5)': ['IDN', 'MMR', 'BGD', 'PAK', 'PHL'],
 'country_count': 19,
 'time_series': '2016 ~ 2023',
 'observations_bycolumn': mobile_money_transactions    152
 dtype: int64}

For the mobile money transactions dataset, we collected data from 19 countries in South and Southeast Asia spanning the years 2016 to 2023. The indicator used is the total number of mobile money transactions per year, resulting in 152 valid observations. This subset helps us capture recent trends in digital financial inclusion across the region.

## 1.7 Fertilizer use (Nitrogen application per farmland)
- Source: Global data on fertilizer use by crop and by country, DRYAD(https://datadryad.org/dataset/doi:10.5061/dryad.2rbnzs7qh)

# 2. Merge Datasets

In [50]:
import pandas as pd

def merge_and_count(df_1, df_2, how = 'inner'):
    country_col = 'country'
    date_col = 'year'
    rows_df_1, rows_df_2 = len(df_1), len(df_2)
    merged_df = pd.merge(df_1, df_2, how=how, on=[country_col, date_col])

    countries_df_1 = df_1[country_col].unique()
    countries_merged = merged_df[country_col].unique()

    dropped_countries = set(countries_df_1) - set(countries_merged)

    rows_merged = len(merged_df)
    countries_df_1_count, countries_merged_count = (
        len(countries_df_1),
        len(countries_merged)
    )

    print(f"Rows in df_1: {rows_df_1}, Countries in df_1: {countries_df_1_count}")
    print(f"Rows in df_2: {rows_df_2}, Countries in df_2: {countries_merged_count}")
    print(f"Rows in merged DataFrame: {rows_merged}, Countries in merged DataFrame: {countries_merged_count}")
    print(f"Dropped countries from df_1: {list(dropped_countries)}")

    return merged_df

## 2.1 Merge con and agri

In [51]:
df_merge_1 = merge_and_count(
    df_1 = df_con,
    df_2 = df_agri,
    how = 'outer'
)
df_merge_1.head(2)


Rows in df_1: 208, Countries in df_1: 16
Rows in df_2: 151, Countries in df_2: 16
Rows in merged DataFrame: 236, Countries in merged DataFrame: 16
Dropped countries from df_1: []


country  year  Fertility rate, total (births per woman)  \
0     BGD  2011                                     2.309   
1     BGD  2012                                     2.263   

   GDP per capita, PPP (constant 2021 international $)  \
0                                        4352.426434     
1                                        4592.048409     

   Life expectancy at birth, total (years)  \
0                                   68.499   
1                                   68.989   

   Mortality rate, infant (per 1,000 live births)  \
0                                            35.4   
1                                            34.1   

   Population density (people per sq. km of land area)  \
0                                        1179.926834     
1                                        1191.289091     

   Population growth (annual %)  Population, total  \
0                      0.908526        153591076.0   
1                      0.958356        155070101.0   

   Poverty headcount ratio at national poverty lines (% of population)  \
0                                                NaN                     
1                                                NaN                     

   Rural population (% of total population)  \
0                                    68.775   
1                                    68.007   

   Urban population (% of total population)   FarmCredit  ICTPolicy  \
0                                    31.225  3271.100803        NaN   
1                                    31.993  3213.392664        NaN   

   ProductionValue  ProcessingValue  Fertilizer  
0     1.649775e+06              NaN         NaN  
1     1.854820e+06    249808.744409         NaN

## 2.2 Merge df_merge1 and GFI

In [52]:
df_merge_2 = merge_and_count(
    df_1 = df_merge_1,
    df_2 = df_gfi,
    how = 'outer'
)
df_merge_2.head(2)

Rows in df_1: 236, Countries in df_1: 16
Rows in df_2: 204, Countries in df_2: 17
Rows in merged DataFrame: 248, Countries in merged DataFrame: 17
Dropped countries from df_1: []


country  year  Fertility rate, total (births per woman)  \
0     BGD  2011                                     2.309   
1     BGD  2012                                     2.263   

   GDP per capita, PPP (constant 2021 international $)  \
0                                        4352.426434     
1                                        4592.048409     

   Life expectancy at birth, total (years)  \
0                                   68.499   
1                                   68.989   

   Mortality rate, infant (per 1,000 live births)  \
0                                            35.4   
1                                            34.1   

   Population density (people per sq. km of land area)  \
0                                        1179.926834     
1                                        1191.289091     

   Population growth (annual %)  Population, total  \
0                      0.908526        153591076.0   
1                      0.958356        155070101.0   

   Poverty headcount ratio at national poverty lines (% of population)  ...  \
0                                                NaN                    ...   
1                                                NaN                    ...   

   Used a mobile phone or the internet to check account balance(% with a financial institution account, age 15+)  \
0                                              10.63                                                               
1                                              10.63                                                               

   Used a mobile phone or the internet to pay bills, female (% age 15+)  \
0                                               1.54                      
1                                               1.54                      

   Used a mobile phone or the internet to pay bills, male (% age 15+)  \
0                                               3.36                    
1                                               3.36                    

   Used a mobile phone or the internet to pay bills, rural (% age 15+)  \
0                                               8.75                     
1                                               8.75                     

   Used a mobile phone or the internet to pay bills, urban (% age 15+)  \
0                                              11.08                     
1                                              11.08                     

   Used a mobile phone or the internet to send money (% age 15+)  \
0                                              17.58               
1                                              17.58               

   Used a mobile phone or the internet to send money, female (% age 15+)  \
0                                               7.64                       
1                                               7.64                       

   Used a mobile phone or the internet to send money, male (% age 15+)  \
0                                              28.28                     
1                                              28.28                     

   Used a mobile phone or the internet to send money, rural (% age 15+)  \
0                                              19.96                      
1                                              19.96                      

   Used a mobile phone or the internet to send money, urban (% age 15+)  
0                                              16.83                     
1                                              16.83                     

[2 rows x 125 columns]

## 2.3 Merge df_merge2 and staple

In [53]:
df_merge_3 = merge_and_count(
    df_1 = df_merge_2,
    df_2 = df_staple,
    how = 'outer'
)
df_merge_3.head(2)

Rows in df_1: 248, Countries in df_1: 17
Rows in df_2: 208, Countries in df_2: 21
Rows in merged DataFrame: 296, Countries in merged DataFrame: 21
Dropped countries from df_1: []


country  year  Fertility rate, total (births per woman)  \
0     AFG  2011                                       NaN   
1     AFG  2012                                       NaN   

   GDP per capita, PPP (constant 2021 international $)  \
0                                                NaN     
1                                                NaN     

   Life expectancy at birth, total (years)  \
0                                      NaN   
1                                      NaN   

   Mortality rate, infant (per 1,000 live births)  \
0                                             NaN   
1                                             NaN   

   Population density (people per sq. km of land area)  \
0                                                NaN     
1                                                NaN     

   Population growth (annual %)  Population, total  \
0                           NaN                NaN   
1                           NaN                NaN   

   Poverty headcount ratio at national poverty lines (% of population)  ...  \
0                                                NaN                    ...   
1                                                NaN                    ...   

   Used a mobile phone or the internet to pay bills, rural (% age 15+)  \
0                                                NaN                     
1                                                NaN                     

   Used a mobile phone or the internet to pay bills, urban (% age 15+)  \
0                                                NaN                     
1                                                NaN                     

   Used a mobile phone or the internet to send money (% age 15+)  \
0                                                NaN               
1                                                NaN               

   Used a mobile phone or the internet to send money, female (% age 15+)  \
0                                                NaN                       
1                                                NaN                       

   Used a mobile phone or the internet to send money, male (% age 15+)  \
0                                                NaN                     
1                                                NaN                     

   Used a mobile phone or the internet to send money, rural (% age 15+)  \
0                                                NaN                      
1                                                NaN                      

   Used a mobile phone or the internet to send money, urban (% age 15+)  \
0                                                NaN                      
1                                                NaN                      

   Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer  \
0                                             197.29              
1                                             190.31              

   Food supply quantity (kg/capita/yr)_Starchy Roots  foodSupply_stability  
0                                               5.72                   NaN  
1                                               6.40                   NaN  

[2 rows x 128 columns]

## 2.4 Merge df_merge3 and access

In [54]:
df_merge_4 = merge_and_count(
    df_1 = df_merge_3,
    df_2 = df_access,
    how = 'outer'
)
df_merge_4.head(2)


Rows in df_1: 296, Countries in df_1: 21
Rows in df_2: 224, Countries in df_2: 21
Rows in merged DataFrame: 312, Countries in merged DataFrame: 21
Dropped countries from df_1: []


country  year  Fertility rate, total (births per woman)  \
0     AFG  2011                                       NaN   
1     AFG  2012                                       NaN   

   GDP per capita, PPP (constant 2021 international $)  \
0                                                NaN     
1                                                NaN     

   Life expectancy at birth, total (years)  \
0                                      NaN   
1                                      NaN   

   Mortality rate, infant (per 1,000 live births)  \
0                                             NaN   
1                                             NaN   

   Population density (people per sq. km of land area)  \
0                                                NaN     
1                                                NaN     

   Population growth (annual %)  Population, total  \
0                           NaN                NaN   
1                           NaN                NaN   

   Poverty headcount ratio at national poverty lines (% of population)  ...  \
0                                                NaN                    ...   
1                                                NaN                    ...   

   Used a mobile phone or the internet to send money, female (% age 15+)  \
0                                                NaN                       
1                                                NaN                       

   Used a mobile phone or the internet to send money, male (% age 15+)  \
0                                                NaN                     
1                                                NaN                     

   Used a mobile phone or the internet to send money, rural (% age 15+)  \
0                                                NaN                      
1                                                NaN                      

   Used a mobile phone or the internet to send money, urban (% age 15+)  \
0                                                NaN                      
1                                                NaN                      

   Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer  \
0                                             197.29              
1                                             190.31              

   Food supply quantity (kg/capita/yr)_Starchy Roots  foodSupply_stability  \
0                                               5.72                   NaN   
1                                               6.40                   NaN   

   Rural Access to Electricity(Percent of Population)  \
0                                                NaN    
1                                                NaN    

   Mobile Cellular Subscriptions (per 100 people)  \
0                                             NaN   
1                                             NaN   

   Fixed Broadband Subsciptions (per 100 people)  
0                                            NaN  
1                                            NaN  

[2 rows x 131 columns]

## 2.5 Merge df_merge4 and mobile

In [55]:
df_merge_5 = merge_and_count(
    df_1 = df_merge_4,
    df_2 = df_mobile,
    how = 'outer'
)
df_merge_5.head(2)
df_merge_5.to_csv('../Data/processed/merged_5.csv', index=False)

Rows in df_1: 312, Countries in df_1: 21
Rows in df_2: 296, Countries in df_2: 21
Rows in merged DataFrame: 470, Countries in merged DataFrame: 21
Dropped countries from df_1: []


# Part 2

# 3. Descriptive

In [56]:
df_descriptive = df_merge_5.describe(include='all')
df_descriptive
#df_descriptive.to_csv('../reports/merged_all_descriptive.csv')


country         year  Fertility rate, total (births per woman)  \
count      470   470.000000                                358.000000   
unique      21          NaN                                       NaN   
top        IND          NaN                                       NaN   
freq        39          NaN                                       NaN   
mean       NaN  2017.961702                                  2.167215   
std        NaN     3.610961                                  0.606198   
min        NaN  2011.000000                                  0.970000   
25%        NaN  2016.000000                                  1.915000   
50%        NaN  2018.000000                                  2.146500   
75%        NaN  2021.000000                                  2.378000   
max        NaN  2024.000000                                  4.305000   

        GDP per capita, PPP (constant 2021 international $)  \
count                                          356.000000     
unique                                                NaN     
top                                                   NaN     
freq                                                  NaN     
mean                                         20765.975320     
std                                          30076.970409     
min                                           3176.650406     
25%                                           6072.530672     
50%                                           8348.177316     
75%                                          14593.312469     
max                                         132468.930253     

        Life expectancy at birth, total (years)  \
count                                358.000000   
unique                                      NaN   
top                                         NaN   
freq                                        NaN   
mean                                  71.630356   
std                                    4.404532   
min                                   63.859000   
25%                                   68.531500   
50%                                   70.415000   
75%                                   74.898750   
max                                   83.595122   

        Mortality rate, infant (per 1,000 live births)  \
count                                       358.000000   
unique                                             NaN   
top                                                NaN   
freq                                               NaN   
mean                                         24.156425   
std                                          15.166628   
min                                           1.700000   
25%                                           9.400000   
50%                                          22.800000   
75%                                          33.825000   
max                                          71.600000   

        Population density (people per sq. km of land area)  \
count                                          328.000000     
unique                                                NaN     
top                                                   NaN     
freq                                                  NaN     
mean                                           714.062798     
std                                           1785.129589     
min                                             18.610515     
25%                                             86.291089     
50%                                            192.753282     
75%                                            379.317000     
max                                           7965.878492     

        Population growth (annual %)  Population, total  \
count                     358.000000       3.580000e+02   
unique                           NaN                NaN   
top                              NaN                NaN   
freq                             NaN                NaN   
mean       

## Select Variables

Analyze the relationship between Food supply quantity (cereals) as the proxy of agricultural output and mobile cellular subscription as the proxy of mobile network penetration and the population's access to communication infrastructure

## 3.1 Independent Variable (mobile cellular subscription )

In [57]:
print(df_descriptive['Mobile Cellular Subscriptions (per 100 people)'])

count     324.000000
unique           NaN
top              NaN
freq             NaN
mean      110.892398
std        33.200936
min         2.497502
25%        83.685324
50%       115.403021
75%       136.952752
max       181.767026
Name: Mobile Cellular Subscriptions (per 100 people), dtype: float64


The variable **"Mobile Cellular Subscriptions (per 100 people)"** has 951 observations with a mean of 45 people. The values range from 0 to 181 people, indicating a high variation across countries and years. The standard deviation is 55 people. 

## 3.2 Outcome variables

For now, we use 
as a preliminary assessment, 6 variables were selectd 
1) (<u>financial access</u>) Financial institution account, female (% age 15+)
2) (<u>e-commerce</u>) Made a digital in-store merchant payment: using a mobile phone (% age 15+)
3) (<u>mobile transaction</u>) Use a mobile phone or the internet to make payments, buy things, or to send or receive money using a financial institution account (% age 15+)
4) (<u>saving</u>) Saved at a financial institution or using a mobile money account (% age 15+)
5) (<u>Food productivity</u>) Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer
6) (<u>Resilience on food production</u>) rolling_std (let's change variable name as "food supply stability")



In [58]:
print(df_descriptive[['Financial institution account, female (% age 15+)', 'Made a digital in-store merchant payment: using a mobile phone (% age 15+)', 'Saved at a financial institution or using a mobile money account (% age 15+)', 'Store money using a financial institution or a mobile money account (% age 15+)', 'Use a mobile phone or the internet to make payments, buy things, or to send or receive money using a financial institution account (% age 15+)', 'Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer', 'foodSupply_stability']])

        Financial institution account, female (% age 15+)  \
count                                          321.000000   
unique                                                NaN   
top                                                   NaN   
freq                                                  NaN   
mean                                            47.764553   
std                                             26.232088   
min                                              2.950000   
25%                                             28.017500   
50%                                             37.850000   
75%                                             76.640000   
max                                             98.210000   

        Made a digital in-store merchant payment: using a mobile phone (% age 15+)  \
count                                          302.000000                            
unique                                                NaN                            
top      

### Based on the descriptive statistics, we decided to use **"Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer"** as outcome variable for now due its high data availability. 

In [59]:
print(df_descriptive['Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer'])

count     355.000000
unique           NaN
top              NaN
freq             NaN
mean      218.141465
std        42.579252
min       129.680000
25%       184.470000
50%       218.650000
75%       256.775000
max       290.380000
Name: Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer, dtype: float64


The variable **"Food supply quantity (kg/capita/yr)_Cereals – Excluding Beer"** has 374 observations with a mean of approximately 217.90 kg per capita per year. The values range from 129.68 to 290.38, indicating variation in cereal supply across countries and years. The standard deviation is 42.43, showing moderate dispersion from the mean.

## 3.3 Control variables

add demography

In [60]:
print(df_descriptive[['Fertility rate, total (births per woman)', 
                      'GDP per capita, PPP (constant 2021 international $)', 
                      'Life expectancy at birth, total (years)', 
                      'Mortality rate, infant (per 1,000 live births)', 
                      'Population density (people per sq. km of land area)', 
                      'Population growth (annual %)', 
                      'Population, total',
                      'Poverty headcount ratio at national poverty lines (% of population)',
                      'Rural population (% of total population)',
                      'Urban population (% of total population)'
                      ]])

        Fertility rate, total (births per woman)  \
count                                 358.000000   
unique                                       NaN   
top                                          NaN   
freq                                         NaN   
mean                                    2.167215   
std                                     0.606198   
min                                     0.970000   
25%                                     1.915000   
50%                                     2.146500   
75%                                     2.378000   
max                                     4.305000   

        GDP per capita, PPP (constant 2021 international $)  \
count                                          356.000000     
unique                                                NaN     
top                                                   NaN     
freq                                                  NaN     
mean                                         20765.975320   

# 4. Graph (Please see the [Visualization_one](https://github.com/Graspp-25-Spring/graspp_2025s_fintech/blob/main/notebooks/Visualizations_One.ipynb) for ALL the visualization and regression)

In [61]:
import seaborn as sns
sns.scatterplot(
    data = df_merge_5,
    x = 'Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer',
    y = 'Mobile Cellular Subscriptions (per 100 people)'
)

ModuleNotFoundError: No module named 'seaborn'

We can see that in general there's a positive correlation between Mobile Cellular Subscriptions and Food Supply Quantity, despite some outliers exist. 

# 5. Regression

## 5.1 Transform　（updated) -- why only 2020?

In [ ]:
df_feat = df_merge_5.query("year == 2020")[[
    'Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer',
    'Mobile Cellular Subscriptions (per 100 people)'
]]
df_feat.head(5)

df_feat_with_control = df_merge_5.query("year == 2020")[[
    'Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer',
    'Mobile Cellular Subscriptions (per 100 people)',
    'Fertility rate, total (births per woman)', 
    'GDP per capita, PPP (constant 2021 international $)', 
    'Life expectancy at birth, total (years)', 
    'Mortality rate, infant (per 1,000 live births)', 
    'Population density (people per sq. km of land area)', 
    'Population growth (annual %)', 
    'Population, total',
    'Poverty headcount ratio at national poverty lines (% of population)',
    'Rural population (% of total population)',
    'Urban population (% of total population)'
]]

df_feat_with_control.head(5)


Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer  \
13                                             189.06              
14                                             189.06              
37                                             278.99              
38                                             278.99              
39                                             278.99              

    Mobile Cellular Subscriptions (per 100 people)  \
13                                             NaN   
14                                             NaN   
37                                      105.291163   
38                                      105.291163   
39                                      105.291163   

    Fertility rate, total (births per woman)  \
13                                       NaN   
14                                       NaN   
37                                     2.175   
38                                     2.175   
39                                     2.175   

    GDP per capita, PPP (constant 2021 international $)  \
13                                                NaN     
14                                                NaN     
37                                        7015.197928     
38                                        7015.197928     
39                                        7015.197928     

    Life expectancy at birth, total (years)  \
13                                      NaN   
14                                      NaN   
37                                   71.421   
38                                   71.421   
39                                   71.421   

    Mortality rate, infant (per 1,000 live births)  \
13                                             NaN   
14                                             NaN   
37                                            24.9   
38                                            24.9   
39                                            24.9   

    Population density (people per sq. km of land area)  \
13                                                NaN     
14                                                NaN     
37                                        1277.544934     
38                                        1277.544934     
39                                        1277.544934     

    Population growth (annual %)  Population, total  \
13                           NaN                NaN   
14                           NaN                NaN   
37                      0.836311        166298024.0   
38                      0.836311        166298024.0   
39                      0.836311        166298024.0   

    Poverty headcount ratio at national poverty lines (% of population)  \
13                                                NaN                     
14                                                NaN                     
37                                                NaN                     
38                                                NaN                     
39                                                NaN                     

    Rural population (% of total population)  \
13                                       NaN   
14                                       NaN   
37                                    61.823   
38                                    61.823   
39                                    61.823   

    Urban population (% of total population)  
13                                       NaN  
14                                       NaN  
37                                    38.177  
38                                    38.177  
39                                    38.177

In [ ]:
import numpy as np
df_feat = df_feat.assign(
    food_log = df_feat['Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer'].apply(np.log),
    mobile_log = df_feat['Mobile Cellular Subscriptions (per 100 people)'].apply(np.log)
)
df_feat.head(5)

NameError: name 'df_feat' is not defined

## 5.1.1 Transform (year 2011 -- 2023)

In [ ]:
import numpy as np

df_reg = df_merge_5.assign(
    food_log = df_merge_5['Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer'].apply(np.log),
    mobile_log = df_merge_5['Mobile Cellular Subscriptions (per 100 people)'].apply(np.log)
)
df_reg.head(5)



country  year  Fertility rate, total (births per woman)  \
0     AFG  2011                                       NaN   
1     AFG  2012                                       NaN   
2     AFG  2013                                       NaN   
3     AFG  2014                                       NaN   
4     AFG  2015                                       NaN   

   GDP per capita, PPP (constant 2021 international $)  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   Life expectancy at birth, total (years)  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

   Mortality rate, infant (per 1,000 live births)  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   

   Population density (people per sq. km of land area)  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   Population growth (annual %)  Population, total  \
0                           NaN                NaN   
1                           NaN                NaN   
2                           NaN                NaN   
3                           NaN                NaN   
4                           NaN                NaN   

   Poverty headcount ratio at national poverty lines (% of population)  ...  \
0                                                NaN                    ...   
1                                                NaN                    ...   
2                                                NaN                    ...   
3                                                NaN                    ...   
4                                                NaN                    ...   

   Used a mobile phone or the internet to send money, urban (% age 15+)  \
0                                                NaN                      
1                                                NaN                      
2                                                NaN                      
3                                                NaN                      
4                                                NaN                      

   Food supply quantity (kg/capita/yr)_Cereals - Excluding Beer  \
0                                             197.29              
1                                             190.31              
2                                             193.09              
3                                             190.09              
4                                             195.92              

   Food supply quantity (kg/capita/yr)_Starchy Roots  foodSupply_stability  \
0                                               5.72                   NaN   
1                                               6.40                   NaN   
2                                              14.55                   NaN   
3                                              13.46              5.345682   
4                                              13.96              3.242869   

   Rural Access to Electricity(Percent of Population)  \
0                                                NaN    
1                                                NaN    
2                                              

## 5.2 Regression

In [ ]:
%pip install statsmodels

import statsmodels.api as sm


def regression(df, x, y):
    cols = x + [y]
    data = df[cols].dropna()
    X = data[x]
    X = sm.add_constant(X) # Capital X for convention
    Y = data[y] # Capital Y for convention

    # Create and fit the OLS model
    model = sm.OLS(Y, X)
    results = model.fit()

    return results.summary()


regression(
    df=df_feat,
    x=['mobile_log'],
    y='food_log'
)


[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


NameError: name 'df_feat' is not defined

There is a positive elasticity between mobile cellular subscriptions and food supply (measured as cereal supply per capita). A 10% increase in mobile subscriptions is associated with roughly a 2.7% increase in food supply, all else equal. However, the relationship is only marginally statistically significant, and the model explains a modest portion of the variation in food supply across countries.

## Regression with control variables 

In [ ]:
import statsmodels.api as sm

def ols(df, y_data, y_feat, x_data, x_feat, controls=None):
    y_col = f"{y_data}_{y_feat}" if y_feat else y_data
    x_col = f"{x_data}_{x_feat}" if x_feat else x_data

    # Include controls in column selection if provided
    cols = [y_col, x_col] + (controls if controls else [])
    data = df[cols].dropna()

    y = data[y_col]
    X = data[[x_col] + (controls if controls else [])]
    X = sm.add_constant(X)  # Add intercept

    model = sm.OLS(y, X)
    results = model.fit()
    print(results.summary())


ModuleNotFoundError: No module named 'statsmodels'

In [ ]:
controls=[
        'Fertility rate, total (births per woman)', 
                      'GDP per capita, PPP (constant 2021 international $)', 
                      'Life expectancy at birth, total (years)', 
                      'Mortality rate, infant (per 1,000 live births)', 
                      'Population density (people per sq. km of land area)', 
                      'Population growth (annual %)', 
                      'Population, total',
                      'Poverty headcount ratio at national poverty lines (% of population)',
                      'Rural population (% of total population)',
                      'Urban population (% of total population)'
    ]

In [ ]:
ols(
    df=df_feat_with_control,
    y_data='food',
    y_feat='log',  # apply log to ∏∏
    x_data='mobile',
    x_feat='log',  # apply log to X
   controls=controls
)


## 5.2.1 Regression(2011-2023)

In [ ]:
ols(
    df=df_reg,
    y_data='food',
    y_feat='log',  # apply log to ∏∏
    x_data='mobile',
    x_feat='log',  # apply log to X)
)